ADBDEV-4548: Implement vacuum/analyze ordering functions for arenadata_toolkit

RekGRpth opened the pull request:
Implement vacuum/analyze ordering functions for arenadata_toolkit

This patch extends the functionality of the arenadata_toolkit extension
by adding two functions to obtain different table orders for vacuum/analyze.
These two functions are planned to be used in the vacuum command in
the operation.py script. By default, the bundle will use the current approach,
in which new tables are placed at the beginning of the list. But users will
be able to reconfigure it to use a second function that places new tables
at the end of the list, or even use a custom query to determine the order.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57657

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/749288

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/749289

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57670

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57671

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/750550

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/750568

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/750551

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/750569

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/750562

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/750548

RekGRpth committed:
Implement vacuum/analyze ordering functions for arenadata_toolkit

This patch extends the functionality of the arenadata_toolkit extension
by adding two functions to obtain different table orders for vacuum/analyze.
These two functions are planned to be used in the vacuum command in
the operation.py script. By default, the bundle will use the current approach,
in which new tables are placed at the beginning of the list. But users will
be able to reconfigure it to use a second function that places new tables
at the end of the list, or even use a custom query to determine the order.

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57690

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/751368

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/751369

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/751434

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/751459

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/751463

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/751562

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql:
```diff
@@ -0,0 +1,55 @@
+$$
+BEGIN
+	RETURN query
+    SELECT n.nspname AS table_schema, c.relname AS table_name
```
There is tab identation in 12th row, but in 39th there are spaces.

RekGRpth replied:
fixed

RekGRpth committed:
space to tab

RekGRpth committed:
Merge branch 'adb-6.x-dev' into ADBDEV-4548

bimboterminator1 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql:
```diff
@@ -0,0 +1,55 @@
+	SELECT n.nspname AS table_schema, c.relname AS table_name
+	FROM pg_class c
+	JOIN pg_namespace n ON c.relnamespace = n.oid
+	LEFT JOIN pg_stat_last_operation o ON o.classid = 'pg_class'::regclass::oid AND o.objid = c.oid AND o.staactionname = actionname
```
`o.staactionname = actionname`. This does not work when `actionname` is lowercase:
```
SELECT n.nspname AS table_schema, c.relname AS table_name, o.staactionname
FROM pg_class c
JOIN pg_namespace n ON c.relnamespace = n.oid AND c.relname='vacuumed'
LEFT JOIN pg_stat_last_operation o ON o.classid = 'pg_class'::regclass::oid AND o.objid = c.oid 
AND o.staactionname = 'VACUUM';

 table_schema | table_name | staactionname 
--------------+------------+---------------
 test_vacuum  | vacuumed   | VACUUM
(1 row)

SELECT n.nspname AS table_schema, c.relname AS table_name, o.staactionname
FROM pg_class c
JOIN pg_namespace n ON c.relnamespace = n.oid AND c.relname='vacuumed'
LEFT JOIN pg_stat_last_operation o ON o.classid = 'pg_class'::regclass::oid AND o.objid = c.oid 
AND o.staactionname = 'vacuum';
 table_schema | table_name | staactionname 
--------------+------------+---------------
 test_vacuum  | vacuumed   | 
(1 row)
```

RekGRpth replied:
> This does not work when `actionname` is lowercase

Yes, as in original query

bimboterminator1 replied:
I.E. it's the expected behaviour?

RekGRpth replied:
I'll clarify this...

RekGRpth replied:
Fixed

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/57839

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/757503

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/758324

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/757504

RekGRpth committed:
add support any case for argument

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/763845

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/763846

InnerLife0 commented on gpcontrib/arenadata_toolkit/sql/adb_vacuum_strategy_test.sql:
```diff
@@ -0,0 +1,67 @@
+
+CREATE TABLE test_vacuum.vacuumed (a int) DISTRIBUTED BY (a);
+CREATE TABLE test_vacuum.not_vacuumed (a int) DISTRIBUTED BY (a);
+SET client_min_messages=WARNING;
```
I think we need a comment about why we want to suppress log messages of create table command.

RekGRpth replied:
added

InnerLife0 commented on gpcontrib/arenadata_toolkit/sql/adb_vacuum_strategy_test.sql:
```diff
@@ -0,0 +1,67 @@
+
+VACUUM test_vacuum.vacuumed;
+
+-- original query from operation.py
```
Do our goal is to move from this query? 
If yes, why do we place it here? It's enough just to show that new function works according to specification.
If no, how will you synchronize it with one in operation.py?

RekGRpth replied:
OK, removed

InnerLife0 commented on gpcontrib/arenadata_toolkit/sql/adb_vacuum_strategy_test.sql:
```diff
@@ -0,0 +1,67 @@
+
+-- default strategy is same as original query
+SELECT * FROM arenadata_toolkit.adb_vacuum_strategy_newest_first('VACUUM') WHERE table_schema = 'test_vacuum';
+-- revert strategy
```
`revert` -> `reversed`

RekGRpth replied:
fixed

InnerLife0 commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql:
```diff
@@ -0,0 +1,55 @@
+/* gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql */
+
+/*
+ * Returns column pair (table_schema, table_name) ordered by haven't been vacuumed
```
It's very hard to read and understand this comment.
Can you try to split and rearrange it and the same one below?

RekGRpth replied:
done

RekGRpth committed:
split and rearrange comments

RekGRpth committed:
comment

RekGRpth committed:
remove original query

RekGRpth committed:
typo

RekGRpth committed:
test

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/58151

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/776601

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/776602

InnerLife0 commented:
As agreed in voice chat, we should create one root function with parameterized query to avoid copypaste.

RekGRpth committed:
optimize

RekGRpth commented:
> As agreed in voice chat, we should create one root function with parameterized query to avoid copypaste.

done

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/58177

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/777936

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/777937

InnerLife0 commented:
> done

Why not create a single plpgsql function with `return query execute` and two simple sql functions, which are well optimized?

RekGRpth commented:
> > done
> 
> Why not create a single plpgsql function with `return query execute` and two simple sql functions, which are well optimized?

`execute` is a bad idea, and plpgsql function is more performant than sql, because of preparing. Current variant is the most performant.

RekGRpth commented:
> two simple sql functions

I may do so
```diff
diff --git a/gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql b/gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql
index e35c904857..e8daaf5839 100644
--- a/gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql
+++ b/gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql
@@ -34,13 +34,10 @@ REVOKE ALL ON FUNCTION arenadata_toolkit.adb_vacuum_strategy_unordered(actionnam
 CREATE FUNCTION arenadata_toolkit.adb_vacuum_strategy_newest_first(actionname TEXT)
 RETURNS TABLE (table_schema NAME, table_name NAME) AS
 $$
-BEGIN
-	RETURN query
 	SELECT u.table_schema, u.table_name
 	FROM arenadata_toolkit.adb_vacuum_strategy_unordered(actionname) u
 	ORDER BY u.statime ASC NULLS FIRST;
-END;
-$$ LANGUAGE plpgsql EXECUTE ON MASTER;
+$$ LANGUAGE sql EXECUTE ON MASTER;
 
 /*
  * Only for admin usage.
@@ -55,13 +52,10 @@ REVOKE ALL ON FUNCTION arenadata_toolkit.adb_vacuum_strategy_newest_first(action
 CREATE FUNCTION arenadata_toolkit.adb_vacuum_strategy_newest_last(actionname TEXT)
 RETURNS TABLE (table_schema NAME, table_name NAME) AS
 $$
-BEGIN
-	RETURN query
 	SELECT u.table_schema, u.table_name
 	FROM arenadata_toolkit.adb_vacuum_strategy_unordered(actionname) u
 	ORDER BY u.statime ASC NULLS LAST;
-END;
-$$ LANGUAGE plpgsql EXECUTE ON MASTER;
+$$ LANGUAGE sql EXECUTE ON MASTER;
 
 /*
  * Only for admin usage.
```

RekGRpth committed:
optimize

RekGRpth commented:
> Why not create a single plpgsql function with `return query execute` and two simple sql functions, which are well optimized?

done

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/58296

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/781378

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/781379

andr-sokolov commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql:
```diff
@@ -0,0 +1,64 @@
+ * and already vacuumed - at the end, else (arg is false) tables that are already
+ * vacuumed are located first, and tables that are not yet vacuumed are located at the end.
+ */
+CREATE FUNCTION arenadata_toolkit.adb_vacuum_strategy_arg(actionname TEXT, arg BOOLEAN)
```
what does it mean "arg"? Can we give more informative name for this argument?

RekGRpth replied:
renamed

andr-sokolov commented on gpcontrib/arenadata_toolkit/arenadata_toolkit--1.2--1.3.sql:
```diff
@@ -0,0 +1,64 @@
+		AND c.relstorage != 'x'
+		AND n.nspname NOT IN (SELECT schema_name FROM arenadata_toolkit.operation_exclude)
+		AND p.parchildrelid IS NULL
+		ORDER BY statime ASC NULLS %s
```
"ORDER BY" should be at the same level as WHERE, because it is not a part of WHERE

RekGRpth replied:
moved

RekGRpth committed:
small cosmetics fixes

RekGRpth committed:
simplify

RekGRpth committed:
explicit pg_catalog

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/792095

RekGRpth committed:
optimize

RekGRpth committed:
fix

BenderArenadata commented:
Allure report https://allure-ee.adsw.io/launch/58497

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/792308

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/792309

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/792306
